runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -7609,11 +7625,15 @@ reservationLockExit(TR::Node *node, int32_t lwOffset, TR::Register *objectClassR
       generateRXInstruction(cg, storeOp, node, monitorReg, generateS390MemoryReference(objReg, lwOffset, cg));
 
       if (!isPrimitive)
+         {   
```

Comment:
In OOL of MonExit, for primitive lock reservation, the above lockward update (Increment to 1 ) in mon exit is done for the helper call. We do not need to update ownedMonitorCount here as in case of primitive we anyway call the helper. So we should be generating the ownedMonitorCount update in OOL in case of non primitive only,

Reply:
If you call to the assembly helpers, they will manage the count for you.

Reply:
Yes, so the JIT code currently in this path calls helper if for primitive lock reservation. For non primitive in this case we have a success so skip helper call. The reason we store the updated lock ward before calling the JIT helper in primitive case, is that, while generating a code for primitive lock reservation for MonEnt, we skip updating Recursion Count, but in case of MonExit, if `OBJECT_HEADER_LOCK_RECURSION_MASK` is set on lockward, it calls helper and to compensate of not incrementing it during monent, we do it here in monexit OOL before calling helper. 

I also had a brief discussion about this with @IBMJimmyk , primitive lock reservation code generated by JIT seems to be incorrect in general. On Z, we would anyway won't generate the lock reservation code as it is disabled due to regression and deadlock seen due to it on couple of workloads.